src/python/pants/backend/scala/lint/scalafmt/rules.py

```diff
@@ -0,0 +1,271 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import os.path
+import textwrap
+from collections import defaultdict
+from dataclasses import dataclass
+
+from pants.backend.scala.lint.scala_lang_fmt import ScalaLangFmtRequest
+from pants.backend.scala.lint.scalafmt.skip_field import SkipScalafmtField
+from pants.backend.scala.lint.scalafmt.subsystem import ScalafmtSubsystem
+from pants.backend.scala.target_types import ScalaSourceField
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.goals.tailor import group_by_dir
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, MergeDigests, PathGlobs, Snapshot
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import BashBinary, FallibleProcessResult, Process
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.testutil.rule_runner import logging
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+_SCALAFMT_CONF_FILENAME = ".scalafmt.conf"
+
+
+@dataclass(frozen=True)
+class ScalafmtFieldSet(FieldSet):
+    required_fields = (ScalaSourceField,)
+
+    source: ScalaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipScalafmtField).value
+
+
+class ScalafmtRequest(ScalaLangFmtRequest, LintRequest):
+    field_set_type = ScalafmtFieldSet
+
+
+class ScalafmtToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = ScalafmtSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: ScalafmtRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Partition:
+    process: Process
+    description: str
+
+
+@dataclass(frozen=True)
+class Setup:
+    partitions: tuple[Partition, ...]
+    original_digest: Digest
+
+
+@dataclass(frozen=True)
+class GatherScalafmtConfigFilesRequest:
+    snapshot: Snapshot
+
+
+@dataclass(frozen=True)
+class ScalafmtConfigFiles:
+    snapshot: Snapshot
+    config_files_for_source_dir: FrozenDict[str, str]
```

Comment:
How about calling it `source_dir_to_config_file`?

Reply:
Sounds fine to me. Will switch.

Reply:
Done.